ecommerce_shop_project/shop/templates/shop/category_products.html

```diff
@@ -37,7 +37,7 @@
         {% for product in products %}
         <div class="col-lg-4 col-md-6 mb-4">
             <div class="card h-100">
-                <a href="#"><img class="card-img-top" src="{{product.image.url}}" alt="" width="250" height="380"></a>
+                <a href="{{ product.get_absolute_url }}"><img class="card-img-top" src="{{product.images.first.image.url}}" alt="" width="250" height="380"></a>
```

Comment:
now if a product doesn't have any images the `first()` returns `none` and template tries call `.image.url` on `none` value. It will obviously fail. But because of template nature it will fail silently but will cause ivalid  HTML `<img>` tag. 

Maybe you can use variable assignment inside iteration what roughly can look like
```html
{% for product in products %}
    {% with cover=product.images.first %}
        {% if cover %}
             <img src="{{ cover.image.url }}" />
        {% else %}
           // potentially some fallback placeholder image if you want :P
            <img src="https://imgflip.com/i/31a3mt" />
        {% endif %}
    {% endwith %}
```

see more https://docs.djangoproject.com/en/2.2/ref/templates/builtins/#with

Reply:
ok, that looks good